omx/gstomxcamera.c

```diff
@@ -0,0 +1,1214 @@
+    gst_object_unref (self->outpool);
+    self->outpool = NULL;
+  }
+
+  if (self->comp) {
+    gst_omx_component_free (self->comp);
+    self->comp = NULL;
+    self->outport = NULL;
+  }
+
+  GST_INFO_OBJECT (self, "Closed component %s", klass->cdata.component_name);
+  return TRUE;
+}
+
+
+static GstStateChangeReturn
+gst_omx_camera_change_state (GstElement * element, GstStateChange transition)
+{
+  GstStateChangeReturn ret = GST_STATE_CHANGE_SUCCESS;
+  GstOMXCamera *self = GST_OMX_CAMERA (element);
+
+  switch (transition) {
+    case GST_STATE_CHANGE_NULL_TO_READY:
+      /* open the component */
+      if (!gst_omx_camera_open (self))
+        return GST_STATE_CHANGE_FAILURE;
+      break;
+    default:
+      break;
+  }
+
+  ret = GST_ELEMENT_CLASS (parent_class)->change_state (element, transition);
+
+  switch (transition) {
+    case GST_STATE_CHANGE_READY_TO_NULL:
+      /* close the component */
+      if (!gst_omx_camera_close (self))
+        return GST_STATE_CHANGE_FAILURE;
+
+      break;
+    default:
+      break;
+  }
+
+  return ret;
+}
+
+static GstFlowReturn
+gst_omx_camera_get_buffer (GstOMXCamera * self, GstBuffer ** outbuf)
+{
+  GstOMXPort *port = self->outport;
+  GstOMXBuffer *buf = NULL;
+  GstOMXAcquireBufferReturn acq_return;
+  OMX_ERRORTYPE err;
+  GstBufferPool *pool = self->outpool;
+  GstFlowReturn flow_ret;
+
+  acq_return = gst_omx_port_acquire_buffer (port, &buf);
+  if (acq_return == GST_OMX_ACQUIRE_BUFFER_ERROR) {
+    goto component_error;
+  } else if (acq_return == GST_OMX_ACQUIRE_BUFFER_FLUSHING) {
+    goto flushing;
+  } else if (acq_return != GST_OMX_ACQUIRE_BUFFER_OK) {
+    return GST_FLOW_ERROR;
+  }
+
+  if (gst_omx_port_is_flushing (port)) {
+    GST_DEBUG_OBJECT (self, "Flushing");
+    gst_omx_port_release_buffer (port, buf);
+    goto flushing;
+  }
+
+  GST_LOG_OBJECT (self, "Handling buffer: 0x%08x %" G_GUINT64_FORMAT,
+      (guint) buf->omx_buf->nFlags, (guint64) buf->omx_buf->nTimeStamp);
+
+  buf->omx_buf->nFilledLen = self->imagesize;
+  if (buf->omx_buf->nFilledLen > 0) {
+    GstMapInfo map = GST_MAP_INFO_INIT;
+
+    GST_LOG_OBJECT (self, "Handling output data");
+
+    if (self->always_copy) {
+      *outbuf = gst_buffer_new_and_alloc (buf->omx_buf->nFilledLen);
+
+      gst_buffer_map (*outbuf, &map, GST_MAP_WRITE);
+      memcpy (map.data,
+          buf->omx_buf->pBuffer + buf->omx_buf->nOffset,
+          buf->omx_buf->nFilledLen);
+      gst_buffer_unmap (*outbuf, &map);
+
+      err = gst_omx_port_release_buffer (port, buf);
+      if (err != OMX_ErrorNone)
+        goto release_error;
+    } else {
+      GstBufferPoolAcquireParams params = { 0, };
+      gint n = port->buffers->len;
+      GstMemory *mem;
+      gint i;
+
+      for (i = 0; i < n; i++) {
+        GstOMXBuffer *tmp = g_ptr_array_index (port->buffers, i);
+
+        if (tmp == buf)
+          break;
+      }
+      g_assert (i != n);
+
+      GST_OMX_BUFFER_POOL (pool)->current_buffer_index = i;
```

Comment:
I'm not sure if this is a safe operation to do

Reply:
@cgomezv  At first sight looks like is not. But the next method (gst_buffer_pool_acquire_buffer) internally uses g_return_val_if_fail to make a check to the index and the buffer extracted from it.

Reply:
Yes, I agree. Also, looking the rest of code only gst_omx_buffer_pool_alloc_buffer function also writes the current_buffer_index variable, and it makes it only during the call to gst_omx_camera_component_init which is called during the component transition from Loaded to Idle. The calls to gst_omx_camera_get_buffer function happen only in PLAY state, so there's no way that both functions could write current_buffer_index at same or similar times.